contributors/design-proposals/auth/svcacct-token-volume-source.md

```diff
@@ -0,0 +1,138 @@
+party to trust that an arbitrary container on the cluster is who it says it is.
+In addition, infrastructure running on top of Kubernetes needs a simple
+mechanism to communicate with the Kubernetes APIs and to provide more complex
+tooling. Finally, a significant set of security challenges are associated with
+storing service account tokens as secrets in Kubernetes and limiting the methods
+whereby malicious parties can get access to these tokens will reduce the risk of
+platform compromise.
+
+As a platform, Kubernetes should evolve to allow identity management systems to
+provide more powerful workload identity without breaking existing use cases, and
+provide a simple out of the box workload identity that is sufficient to cover
+the requirements of bootstrapping low-level infrastructure running on
+Kubernetes. We expect that other systems to cover the more advanced scenarios,
+and see this effort as necessary glue to allow more powerful systems to succeed.
+
+With this feature, we hope to provide a backwards compatible replacement for
+service account tokens that strengthens the security and improves the
+scalability of the platform.
+
+## Proposal
+
+Kubernetes should:
+
+1.  Allow the automatic service account token injection to be disabled via
+    configuration.
+1.  Allow automatic service account token creation to occur only when a user
+    explicitly requests.
+1.  Implement a ServiceAccountToken volume projection that that maintains a
+    service account token requested by the node from the TokenRequest API.
+
+### Token Volume Projection
+
+A new volume projection will be implemented with an API that closely matches the
+TokenRequest API.
+
+```go
+type ServiceAccountTokenProjection struct {
+  // Audiences are the intendend audiences of the token.
+  Audiences []string
+  // ExpirationSeconds is the requested duration of validity of the service
+  // account token
+  ExpirationSeconds *int64
+  // Path is the relative path of the file to project the token into.
+  Path string
+}
+```
+
+A volume plugin implemented in the kubelet will project a service account token
+sourced from the TokenReview API into volumes created from
+ProjectedVolumeSources. As the token approaches expiration, the kubelet volume
+plugin will proactively rotate the service account token.
+
+To replace the current service account token secrets, we also need to inject the
+clusters CA certificate bundle. Initially we will deploy to data in a configmap
+per-namespace and reference it using a ConfigMapProjection.
+
+This fixes one scalability issue with the current service account token
+deployment model where secret GETs are a large portion of overall apiserver
+traffic. However it doesn't solve the per-namespace replicated CA data which
+needs to be stored in ETCD.
+
+To fix that issue we can implement a PublicConfigMapProjection that exactly
+copies the implementation of ConfigMapProjection but sources config from the
+`kube-public` namespace.
```

Comment:
I *think* that this is the first time that `kube-public` is being used as part of the core controllers.  We introduced it as part of the kubeadm bootstrapping process but it was completely optional. Not sure if other stuff has shown up in there.

It is a little strange to make a namespace special like this.  In fact, `kube-system` isn't all that special above and beyond convention (and RBAC defaults, I think?).  Introducing a PublicConfigMapProjection is probably something that deserves its own design doc.  I'd suggest something like listing the namespace and requiring that the target configmap be readable by anonymous.  That would allow folks to use this in other contexts and avoid naming `kube-public` explicitly.